### DISCUSSION TITLE
Single-Page-App router class as part of library

### DISCUSSION THREAD

**Discussion by WolfgangFahl** [zauberzeug/nicegui#1066] [Ideas / Feature Requests]
### Question

https://github.com/zauberzeug/nicegui/tree/main/examples/single_page_app uses a local Router class.
I wonder whether this is part of the library by now or could / should be made part of it?

**Comment by WolfgangFahl** [comment]
May be the component bazaar idea of https://github.com/justpy-org/justpy/discussions/410 is to be discussed again in this context so that a component such a Router would be placed in the "component bazaar" and that way get proper maintainance/support/documentation/advertising.

**Comment by rodja** [comment]
We really like the router class from the SPA example and think it would be a good addition to the core library (no component bazar needed). We have left it as an example for now because there are several things not thought through at the moment. From the top of my head:

- Is the name well-chosen? We already have an [APIRouter](https://nicegui.io/documentation/page#modularize_with_apirouter) class from FastAPI.
- Is the usability good enough? Who is using this already and can provide some feedback?
- What are other use-cases for this?
- Should we split the content-replacing and url modification (with back/forward nav) into separate APIs?

**Comment by rodja** [comment]
There was has been the attempted pull request https://github.com/zauberzeug/nicegui/pull/1716 to introduce error messages and allow parameters in routes. Both very useful features which are still not available in the SPA example.

**Comment by ghost** [comment]
I am a bit confused between ui.refleshable and SPA sample's Router, from my understanding, both of them clear all elements in the container and then recreate them, SPA sample's Router just do more maintain the browser's history popstate event and spa internal urls. So as my point in another discussion #349 , if I don't care about the spa internal urls I just use ui.refleshable, if I want to split the context into multi sub-contexts and reflesh sub-context I can use ui.refleshable, is it right?

**Reply by rodja** [reply]
Yes, that is true. The SPA example is mainly a demonstration on how to manage the routes.

**Comment by Alyxion** [comment]
I think this is an important base feature and thus filed a [pull request](https://github.com/zauberzeug/nicegui/pull/2811) yesterday which is still work in progress though.

As much as I like @WolfgangFahl 's bazar idea in general I think if you want to do this "right" it should be part of the base library. For example ui.navigate.to() is not interceptable on JavaScript side, so the a bazar version had to somehow "hack" (overwrite) the base method in some dirty way. And then there is an update to NiceGUI which changed the signature but not in the bazar component yet... things would get dirty quite quickly. Also as likely many other people have this requirement and then one knows better than the other etc. etc. things could here start to diverge and then in the end you have five versions of the same feature.

Regarding @rodja 's questions:

* Is the name well-chosen? We already have an [APIRouter](https://nicegui.io/documentation/page#modularize_with_apirouter) class from FastAPI.
    * I named it **SinglePageRouter**. Open for everything.
* Is the usability good enough? Who is using this already and can provide some feedback?
   * It was a good  starting point for me for the PR. But
       * it is buggy (the back button does not work, at least you can never return to /)
       * it does not allow embedding the page in a template (e.g. constant headers and footers which do not get exchanged)
       * ui.navigate.to does not know it (and thus is not usable)
       * it does not support hyperlinks
       * it requires "dirty" nesting of all pages within one method context instead of being able to define them as before (as @pages) in a global context
* What are other use-cases for this?
    * Many mentioned in the PR already. Constant connection to the same server and process, lower latency, per-user data can be kept in-memory between page changes allowing "living" Python objects, no script reloads between pages, thus super smooth navigation etc. etc.
* Should we split the content-replacing and url modification (with back/forward nav) into separate APIs?
    * At least worth a thought. I am personally anyway no big fan of binding every page, sub-tab etc. to an explicit URL though its of course nice to have it for some uses cases, e.g. a forum like this where you want to keep the direct link. But in more complex applications... imagine a VS Code in the browser, Office 365 etc. you potentially do not want the history or URL to change at all.
    
Regarding the [#1716](https://github.com/zauberzeug/nicegui/pull/1716): The idea should definitely be picked up, e.g. /mails/123 should definitely be possible in my opinion when implementing an SPA.

Clone the fork / branch https://github.com/Alyxion/nicegui/tree/feature/client_data  and give [examples/single_page_router/advanced.py](https://github.com/Alyxion/nicegui/blob/feature/client_data/examples/single_page_router/advanced.py) a try - feels super smooth.

**Comment by hhstore** [comment]
- `SPA` is an important feature, and would love to be added to `nicegui`. 
- With it, it's easier to package NiceGUI into a desktop app, similar to Tauri.
- I've had multiple nicegui-based projects that use `SPA` so they don't need to force a page refresh.
- I'd love to see nicegui build this feature in.

**Comment by rodja** [comment]
The pull request #2811 is progressing nicely. But the final API is still to be decided. Any input is much appreciated. These things are hard to change once they are established.

**Reply by hhstore** [reply]
## API Design:

- add a new api, like `ui.view()`
- or, extend the old `ui.page()` method by adding `parameters`.


```python

# new api:
ui.page()  # Page Mode
ui.view()  # SPA View Mode


# new arg:
r = APIRouter()
r.page("#view1")               # style 1: by url prefix, like: ["#view1", "#/view2"]
r.page("#/view2", mode="spa")  # style 2
r.page("#view3", is_spa=True)  # style 3
r.view("/view4")               # style 4

```

**Reply by rodja** [reply]
@hhstore thanks for the input. Have you read the discussions in #2811? I like `.view` but your suggestions seem quite unrelated to the API topics which have been discussed on the pull request.

**Comment by vickorian** [comment]
@rodja

Say page 1 has some application running that then needs to spin off a sub application, like say you're working on something and need to spin off an authentication to request some data separate from what you're working on currently. I think that would be a good use for the decorator like you were suggesting in #2811.

 If you could segment the data to allow for communication between pages, but not have persistent memory then you could login to another database and make a request for it while not having to worry about the session. If you don't want to configure the data later or you already know the format and can write code to format it then having a specific tag to reference data in a page and make only that variable available to the general session could be interesting.

```
@spa1.page('/').variable("extracted_data")
def spa1_index():
   ...
```

in that world, I see the decorator as a really powerful stylistic tool instead of having to set a bunch of variables after the fact or ahead of time.

**Reply by rodja** [reply]
Sorry, I do not understand your example `@spa1.page('/').variable("extracted_data")`. Is "extracted_data" an entry in some storage? Why not access it via `app.storage`?

**Reply by vickorian** [reply]
it was my understanding that the app.storage was relatively lightweight, specifically a serializable dict. The datatypes are arbitrary, but i've had bad experiences attempting to manipulate large files or dataframes via dicts if they have to be stored and retrieved. That's why i was suggesting that having a variable to be able to be "seen" by the host process/app would be worthwhile.

**Reply by rodja** [reply]
Ah ok. In 1.4.21 we are introducing `app.storage.client` and `app.storage.tab` which both are not serialized. The whole problem is also related to #2841.

**Comment by WolfgangFahl** [comment]
@hhstore 
While at it i would move away from the singleton ui.call() approach and introduce actual objects:
E.g.
- server
- solution
- page
- view (e.g. tab)

For refrence see:
![grafik](https://github.com/zauberzeug/nicegui/assets/1336221/220d9c23-f8e6-4063-b02a-34c29ac6bf84)


* http://diagrams.bitplan.com/render/png/0x71d4eac9.png
* http://diagrams.bitplan.com/render/txt/0x71d4eac9.txt

https://github.com/WolfgangFahl/nicegui_widgets/blob/main/ngwidgets/webserver.py already follows this approach and e.g. uses a config class for configuration.

**Comment by WolfgangFahl** [comment]
@vickorian 
serizaliazability is also something you'll find at https://github.com/WolfgangFahl/nicegui_widgets/blob/main/ngwidgets/yamlable.py

with a @lod_storable decorator

this is for limited complexity situations where a list of dicts or similar needs to be stored which can be readily transformed to JSON, YAML or SQL/SPARQL. Complex graphs are a different story but calling for a standard here in the future might help - unfortunately there are competing versions such as pydantic, or SQLModel out there which are incompatible. So doing an API support is IMHO better than actively asking for a certains structure - still we might want to be able to show off how things work and provide examples.
